From da293fd9cbfea18c301f2e5d82b70c40196f4295 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Mon, 2 Jun 2014 17:08:21 -0400
Subject: [md] dm-cache: always split discards on cache block boundaries

Message-id: <1401728901-921-4-git-send-email-snitzer@redhat.com>
Patchwork-id: 81836
O-Subject: [RHEL7.1 PATCH 3/3] dm cache: always split discards on cache block boundaries
Bugzilla: 1103836
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1103836

Upstream commit f1daa838e861ae1a0fb7cd9721a21258430fcc8c
Author: Heinz Mauelshagen <heinzm@redhat.com>
Date:   Fri May 23 14:10:01 2014 -0400

    dm cache: always split discards on cache block boundaries

    The DM cache target cannot cope with discards that span multiple cache
    blocks, so each discard bio that spans more than one cache block must
    get split by the DM core.

    Signed-off-by: Heinz Mauelshagen <heinzm@redhat.com>
    Acked-by: Joe Thornber <ejt@redhat.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>
    Cc: stable@vger.kernel.org # v3.9+

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-cache-target.c b/drivers/md/dm-cache-target.c
index eb0e001..bb57299 100644
--- a/drivers/md/dm-cache-target.c
+++ b/drivers/md/dm-cache-target.c
@@ -2172,6 +2172,8 @@ static int cache_create(struct cache_args *ca, struct cache **result)
  ti->num_discard_bios = 1;
  ti->discards_supported = true;
  ti->discard_zeroes_data_unsupported = true;
+ /* Discard bios must be split on a block boundary */
+ ti->split_discard_bios = true;
 
  cache->features = ca->features;
  ti->per_bio_data_size = get_per_bio_data_size(cache);
-- 
1.7.1